<a href="https://colab.research.google.com/github/adarsh2313/ResNet18-Pytorch/blob/main/main.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import torch
import torch.nn as nn
import torch.optim as optim
device = torch.device("cuda")

In [3]:
from resnet import model
from dataset import train_dl,test_dl

In [4]:
torch.cuda.empty_cache()
model.cuda()

ResNet18(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer0): block0(
    (conv): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=same, bias=False)
    (bn): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
  (layer1): block(
    (conv_stride2): Conv2d(64, 128, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (bn): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv_stride1): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=same, bias=False)
    (resize): Conv2d(64, 128, kernel_size=(1, 1), stride=(2, 2), bias=False)
  )
  (layer2): block(
    (conv_stride2): Conv2d(128, 256, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (bn): BatchNo

Training functions:


In [5]:
optimizer = optim.Adam(model.parameters(), lr = 0.01)
lossfn = nn.CrossEntropyLoss()
epochs = 50

In [7]:
for epoch in range(epochs):
  
  epoch_loss,epoch_acc = 0,0
  model.train()

  for input,label in train_dl:

    input = input.cuda()
    label = label.cuda()

    pred = model(input)
    loss = lossfn(pred,label)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    acc = ((pred.argmax(dim=1) == label).float().mean())
    epoch_acc += acc/len(train_dl)
    epoch_loss += loss.item()/len(train_dl)

  if (epoch+1)%5 == 0:
    print("For epoch",epoch+1,": Loss = {:.2f}".format(epoch_loss),"and Accuaracy = {:.2f}".format(float(epoch_acc*100)))

For epoch 5 : Loss = 0.93 and Accuaracy = 64.10
For epoch 10 : Loss = 0.82 and Accuaracy = 67.99
For epoch 15 : Loss = 0.74 and Accuaracy = 72.68
For epoch 20 : Loss = 0.48 and Accuaracy = 83.30
For epoch 25 : Loss = 0.42 and Accuaracy = 85.00
For epoch 30 : Loss = 0.37 and Accuaracy = 87.53
For epoch 35 : Loss = 0.16 and Accuaracy = 94.33
For epoch 40 : Loss = 0.24 and Accuaracy = 91.67
For epoch 45 : Loss = 0.10 and Accuaracy = 96.75
For epoch 50 : Loss = 0.16 and Accuaracy = 95.00


Validation function:


In [37]:
with torch.no_grad():

  epoch_acc = 0
    
  for input,label in test_dl:
      
    input = input.cuda()
    label = label.cuda()

    pred = model(input)
    
    acc = ((pred.argmax(dim=1) == label).float().mean())
    epoch_acc += acc/len(test_dl)
  
print("Accuaracy = {:.2f}".format(float(epoch_acc*100)))

Accuaracy = 19.61


Getting 95% accuracy rate while training but only 19.61% on validation; possible case of model overfitting. Rectify that